### PR TITLE
Add kickstart tests for the image deployment

### PR DESCRIPTION
**Pull request by poncovka**

### CODE REVIEW

**Comment by poncovka** [comment]
The `anaconda-conf` test was fixed. Thanks @jstodola for pointing it out!

File: bare-metal-1.ks.in [hunk 1]
```diff
@@ -0,0 +1,99 @@
+#version=DEVEL
+#test name: bare-metal-1
+
+# Use defaults for network, bootloader and storage.
+%ksappend network/default.ks
+%ksappend bootloader/default.ks
+%ksappend storage/default.ks
+
+# Install the prepared tar.
+liveimg --url=http://10.43.136.2/trees/rv/bmi/BP2-0.0.4.tar.xz
```
Comment (jkonecny12) [inline bare-metal-1.ks.in:10]: I think this URL should be a fragment.

File: bare-metal-1.ks.in [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+diff /tmp/expected.out /tmp/generated.out
+
+if [[ $? != 0 ]]; then
+    echo "*** unexpected files has changed:" >> /mnt/sysroot/root/RESULT
+    cat /tmp/generated.out >> /mnt/sysroot/root/RESULT
+fi
+
+%end
+
+%post --nochroot
```
Comment (jkonecny12) [inline bare-metal-1.ks.in:67]: Would be great to have a comment explaining what you are trying to achieve in the post section.

File: bare-metal-1.ks.in [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+./var/lib/unbound/root.key
+EOF
+
+# Check the output
+diff /tmp/expected.out /tmp/generated.out
+
+if [[ $? != 0 ]]; then
+    echo "*** unexpected files has changed:" >> /mnt/sysroot/root/RESULT
+    cat /tmp/generated.out >> /mnt/sysroot/root/RESULT
+fi
```
Comment (jkonecny12) [inline bare-metal-1.ks.in:63]: Shouldn't we also compare files directly. I mean if file in a tarball is really what we installed? It could be done by hashing or just `diff -q`.

File: bare-metal-2.ks.in [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#version=DEVEL
+#test name: bare-metal-2
+# NOTE: This is not a supported use case, but it is easier to test.
```
Comment (jkonecny12) [inline bare-metal-2.ks.in:3]: Please elaborate a bit more on this. What is the use case here?

**Review by jkonecny12 — CHANGES_REQUESTED** [review]
Looks good to me in general but I'm missing information. It's hard to follow these tests in the current form. Could you please add more comments and explanation about what is happening there?

File: functions.sh [hunk 5]
```diff
@@ -260,3 +265,41 @@ remove_iscsi_target() {
...
+    # Create the updates directory.
+    mkdir -p "${updates_dir}"
+
+    # Check the updates image.
+    if [[ -z "${image_url}" ]]; then
+        return
+    fi
+
+    # Download and extract the updates image.
+    ( cd "${updates_dir}" ; curl -f "${image_url}" | gzip -dc  | cpio -idu )
```
[the remark this answers is not in the document]
Comment (poncovka) [reply functions.sh:282]: The command is using `cd`, but I don't want to change the current directory of the post script.

File: bare-metal-1.ks.in (hunk 3 above)
Comment (poncovka) [reply bare-metal-1.ks.in:63]: That was the plan, but I am not able to unpack the original tar image in stage2. I would have to use an additional storage and that would be painful. I had outputs from the diff as well and the results are the same.

**Review by rvykydal — APPROVED** [review]
Looks great to me, I have just a remark about test naming.
I think that of the feature Bare Metal Image Deployment `image-deployment` would describe the purpose / use case of the test than `bare-metal`.
Yes, I've myself been using "bare metal installation" to refer to the feature even though it does not make sense to me in the context of installer (unlike in the context of Image Builder where it makes more sense)

**Comment by poncovka** [comment]
I will open separate pull requests for some of these changes.

File: functions.sh [hunk 6]
```diff
@@ -260,3 +265,41 @@ remove_iscsi_target() {
...
+upload_updates_image() {
+    local tmp_dir="${1}"
+    local updates_img="${2}"
+
+    # Copy the updates image.
+    mkdir ${tmpdir}/http
+    cp "${updates_img}" ${tmpdir}/http/updates.img
+
+    # Start the http server.
+    start_httpd ${tmpdir}/http $tmpdir
```
[the remark this answers is not in the document]
Comment (poncovka) [reply functions.sh:301]: Fixed in https://github.com/rhinstaller/kickstart-tests/pull/467.

File: functions.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply functions.sh:282]: The `pushd` and `popd` is generally better solution.

[the remark this answers is not in the document]
Comment (poncovka) [reply functions.sh:282]: This line is copy-pasted from the `apply-updates` script. I don't really want to experiment with alternative implementations.

File: bare-metal-1.ks.in (hunk 1 above)
Comment (poncovka) [reply bare-metal-1.ks.in]: I need the URL also in the `%post` script. Anyway, this test is be disabled by default because of the tar image. I have added a comment there to make it clear.

File: bare-metal-1.ks.in (hunk 2 above)
Comment (poncovka) [reply bare-metal-1.ks.in]: Every test has a description at the top of the `ks.in` file now.

File: bare-metal-2.ks.in (hunk 4 above)
Comment (poncovka) [reply bare-metal-2.ks.in]: Fixed.

**Comment by poncovka** [comment]
Rebased and updated.

**Comment by poncovka** [comment]
@jkonecny12, can I ask for a review, please?

**Review by jkonecny12 — APPROVED** [review]
Looks good to me. We just should remove cobra02 ip from this in the future but lets use this solution for now.
